vnext/Scripts/installRNW.js

```diff
@@ -0,0 +1,121 @@
+/**
+ * @format
+ */
+// @ts-check
+const path = require('path');
+const fs = require('fs');
+const glob = require('glob');
+const rimraf = require('rimraf');
+
+function retryOnError(errorCode, func) {
+  const startTime = Date.now();
+
+  while (true) {
+    try {
+      func();
+      return true;
+    } catch (ex) {
+      if (ex.code !== errorCode) {
+        throw ex;
+      }
+    }
+
+    if (Date.now() - startTime > 5000) {
+      return false;
+    }
+  }
+}
+
+function copyDirectories(srcPath, targetPath, dirSpecs) {
+  dirSpecs.forEach(dirSpec => {
+    if (dirSpec.mergeFiles !== true) {
+      // rimraf issue 72: We will see ENOTEMPTY if Windows is still holding a
+      // lock to a file in our directory. E.g. if we get unlucky with timing
+      // and antivirus decides to run. Keep on spinning until we're able to
+      // delete the directory.
+      let deleted = retryOnError('ENOTEMPTY', () => {
+        if (dirSpec.rmFilter === undefined) {
+          rimraf.sync(path.resolve(targetPath, dirSpec.dest));
+        } else {
+          rimraf.sync(
+            path.resolve(targetPath, dirSpec.dest + path.sep + dirSpec.filter),
+          );
+        }
+      });
+
+      if (!deleted) {
+        throw new Error(
+          'Timed out trying to delete directory. Ensure no locks are being held to project files and try again.',
+        );
+      }
+    }
+
+    const curSrcPath = path.resolve(srcPath, dirSpec.src);
+    const curTargetPath = path.resolve(targetPath, dirSpec.dest);
+
+    const globSpec = dirSpec.rmFilter
+      ? '**/' + dirSpec.rmFilter
+      : '**/*.{js,png,gif,h}';
+
+    glob.sync(globSpec, {cwd: curSrcPath}).forEach(file => {
+      const dir = path.dirname(file);
+      const targetDir = path.resolve(curTargetPath, dir);
+      const targetFile = path.resolve(curTargetPath, file);
+
+      fs.mkdirSync(targetDir, {recursive: true});
+      fs.writeFileSync(
+        targetFile,
+        fs.readFileSync(path.join(curSrcPath, file)),
+      );
+    });
+  });
+}
+
+function copyFile(srcPath, targetPath, filename) {
+  fs.writeFileSync(
+    path.resolve(targetPath, filename),
+    fs.readFileSync(path.resolve(srcPath, filename)),
+  );
+}
+
+
+exports.installRNW = () => {
+  const reactNativePath = path.dirname(
+    require.resolve('react-native/package.json'),
+  );
+
+  const reactNativeWindowsPath = path.dirname(
+    require.resolve('react-native-windows/package.json'),
+  );
+
+  const installationPath = path.resolve(reactNativeWindowsPath, '../../react-native-installation');
```

Comment:
>path.resolve(reactNativeWindowsPath, '../../react-native-installation'); [](start = 27, length = 72)

I'd love to find a cleaner way of doing this.  I'm trying to find the app directory, but it's being done from an install script that runs from react-native-windows.  I can't use process.cwd() as that just returns the react-native-windows directory.  So I'm currently going up two directories from react-native-windows.  I searched around the internet and found others having the same issue, but I haven't found a cleaner solution yet.

Reply:
The only way I can think of of getting the correct directory here is to not do it during RNW's postInstall.  Maybe if we moved it to an ensure step as part of run-windows?  -- And probably also expose it as a separate standalone command.

Basically add a `react-native prepare-windows` command.
Then add a default postInstall script to the apps package.json in the template:
"postInstall": "react-native prepare-windows"

Eventually we'd want to hook up logic to allow all the platforms to be queried as a global `react-native prepare` step, which would ensure all the platforms get hooked up to the same location.

As it is currently, macos and windows might end up using a different location if one but not the other is hoisted in a monorepo, which would prevent the whole thing from working.

Reply:
For now I'm using a simple convention of going up two directories from wherever react-native-windows was installed, and using that convention everywhere.  Another thought:  maybe we could search for react-native.config.js, and whatever ancestor-most directory has one is the directory we install to.

---
In reply to: [405080947](https://github.com/microsoft/react-native-windows/pull/4522#discussion_r405080947) [](ancestors = 405080947)